### ISSUE TITLE
Property getters/setters 'lost' when accessed from interface

### ISSUE THREAD

**Issue by tomblind** [Perryvw/TypescriptToLua#206]
Properties with custom getters/setters in typescript only work in lua if accessed from a variable with the classes type. If accessed from a variable with an interface that maps to the class, the lua tries to access the property directly instead of calling the getter/setter.

Example:
```ts
interface Foo
{
    prop: string;
}

class Bar
{
    private _prop = "BAR";
    public get prop() { return this._prop; }
    public set prop(value: string) { this._prop = value; }
}

let bar = new Bar();
print(bar.prop);

let foo: Foo = bar;
print(foo.prop);
```
Produces:
```lua
Bar = Bar or {}
Bar.__index = Bar
function Bar.new(construct, ...)
    local instance = setmetatable({}, Bar)
    instance._prop = "BAR"
    if construct and Bar.constructor then Bar.constructor(instance, ...) end
    return instance
end
function Bar.constructor(self)
end
function Bar.get__prop(self)
    return self._prop
end
function Bar.set__prop(self,value)
    self._prop = value;
end
local bar = Bar.new(true);
print(bar:get__prop()); --prints "BASE"
local foo = bar;
print(foo.prop); --prints nil
```

One possible solution would be to use a custom metatable for classes with properties like these:
```lua
Bar = Bar or {}
Bar__meta =
{
    getters =
    {
        prop = function(self)
            return self._prop
        end
    },
    setters =
    {
        prop = function(self, value)
            self._prop = value
        end
    },
    __index = function(self, key)
        local getter = Bar__meta.getters[key]
        if getter then
            return getter(self)
        end
        return rawget(self, key) or Bar[key]
    end,
    __newindex = function(self, key, value)
        local setter = Bar__meta.setters[key]
        if setter then
            setter(self, value)
        else
            rawset(self, key, value)
        end
    end
}
function Bar.new(construct, ...)
    local instance = setmetatable({}, Bar__meta)
    instance._prop = "BAR"
    if construct and Bar.constructor then Bar.constructor(instance, ...) end
    return instance
end
function Bar.constructor(self)
end
local bar = Bar.new(true);
print(bar.prop);
local foo = bar;
print(foo.prop);
```

BTW, sorry for all the bugs. I have a large lua project which would really benefit from static typing so I'm putting tstl through it's paces to see if it will work for what I need. :)

**Comment by lolleko** [comment]
Would be great if we could handle this at compile time somehow.
I think we should only use metatable "hacks" as last resort.

**Comment by Perryvw** [comment]
I have to think about this a bit more, but I think we might not have a choice there. Imagine 2 concrete classes implementing the same interface, one using getters and the other without. My current intuition would be there is no way to figure this out at compile time, but like I said I'll think about it a bit more.

**Comment by tomblind** [comment]
Personally, I wouldn't consider it a hack. This is how property bindings to other languages are typically handled. Which is why (I suspect) lua supports assigning custom functions to __index and __newindex. It does add additional overhead, however. It would be best to only use this for classes with getters/setters.

**Comment by lolleko** [comment]
Maybe hack wasn't the correct word, let me rephrase: using meta tables will have impact on the runtime performance. Like I said would be great if we could avoid that.

**Comment by tomblind** [comment]
Now that classes have been updated, I want to come back around to solving this issue. The problem with my original proposed solution is that it breaks when used with inheritance. Consider the following scenario:
```ts
class Foo {
    prop = "foo";
    get foo() { return this.prop; }
}
class Bar extends Foo {
    get bar() { return this.prop + "bar"; }
}
const b = new Bar();
print(b.foo);
```

Using an `__index` function as suggested, would produce something like this:
```lua
Foo.____getters = {
    foo = function(self)
        return self.prop
    end
}
...
Foo.prototype.__index = function(self, key)
    local getter = Foo.____getters[key]
    if getter then
        return getter(self)
    end
    return Foo.prototype[key]
end
...
Bar.____getters = {
    bar = function(self)
        return self.prop .. "bar"
    end
}
...
Bar.prototype.__index = function(self, key)
    local getter = Bar.____getters[key]
    if getter then
        return getter(self)
    end
    return Bar.prototype[key]
end
...
local b = Bar.new()
print(b.foo)
```
The problem here is a bit subtle. When trying to access the getter `foo`, Bar's `__index` function will chain up to Foo's, which will find the getter and call it. But, the value of `self` in Foo's `__index` won't be the original table, but actually `Bar.prototype`! So the result will incorrectly be `nil`.

The fundamental issue is that the original `self` does not get maintained when moving up the prototype chain. There's a couple of ways to work around this, but none are very simple (especially if we want to ensure that there's no additional overhead for classes without getters/setters).

The (arguably) best way to handle this, is to treat the getter/setter tables as separate prototype chains and then give classes that use them a special metatable that first looks up missing properties there.

Example:
```lua
...
Foo.____getters = {
    foo = function(self)
        return self.prop
    end
}
Foo.____getters.__index = Foo.____getters
...
Foo.____meta = {
    __index = function(self, key)
        local getter = Foo.____getters[key]
        if getter then
            return getter(self)
        end
        return Foo.prototype[key]
    end
}
...
Foo.new = function(...)
    self = setmetatable({}, Foo.____meta)
    self:____constructor(...)
    return self
end
...
Bar.____getters = {
    bar = function(self)
        return self.prop .. "bar"
    end
}
Bar.____getters.__index = Bar.____getters
setmetatable(Bar.____getters, Foo.____getters)
...
Bar.____meta = {
    __index = function(self, key)
        local getter = Bar.____getters[key]
        if getter then
            return getter(self)
        end
        return Bar.prototype[key]
    end
}
...
Bar.new = function(...)
    self = setmetatable({}, Bar.____meta)
    self:____constructor(...)
    return self
end
```
The way this works at runtime is that when a property doesn't exist on a table, its custom metatable will first look up the getter chain, then up the standard prototype chain. This example only shows getters, but setters would be handled the same way.

Classes without getters/setters would not receive any of this, of course. But the one caveat is that classes that extend others which have getters/setters would also need to use this system, even if they don't define their own.

Thoughts?

**Comment by tomblind** [comment]
An alternative way of handling this would be to actually take over traversing the prototype chain altogether:
```lua
function __TS_index(self, key)
    local proto = getmetatable(self)
    while proto do
        local val = rawget(proto, key)
        if val ~= nil then
            return val
        end

        local getters = rawget(proto, "____getters")
        if getters then
            local getter = getters[key]
            if getter then
                return getter(self)
            end
        end

        local super = rawget(rawget(proto, "constructor"), "____super")
        if not super then
            break
        end

        proto = rawget(super, "prototype")
    end
end
...
Foo.prototype.__index = __TS_index;
...
Foo.prototype.____getters = {
    foo = function(self)
        return self.prop;
    end;
}
...
Bar.prototype.__index = __TS_index;
...
Bar.prototype.____getters = {
    bar = function(self)
        return self.prop .. "bar";
    end;
}
```

I hesitated to suggest this at first since it seems a bit hacky, but it actually has some good benefits:
- Keeps custom behavior cleanly separated
- Reduces amount of traversal of prototype chains happening (O(N) vs O(2N) in the previous suggestion)
- Uses fewer tables/metatables

But the index function is pretty heavy duty and should probably be relegated to a lib function. I don't know if I love having core class features depend on lib functions.
